src/common/operator/cast_operators.cpp

```diff
@@ -811,6 +811,16 @@ struct IntegerCastOperation {
 		return true;
 	}
 
+	template <class T, bool NEGATIVE>
+	static bool HandleHexDigit(T &state, uint8_t digit) {
+		using result_t = typename T::Result;
+		if (state.result > (NumericLimits<result_t>::Maximum() - digit) / 16) {
```

Comment:
Since this is a power of two, we can avoid doing the (more expensive) division / multiplication, in favor of bit shifts
x / 16 == x >> 4
x * 16 == x << 4

Reply:
Generally the compiler will make this optimization for you if you are dividing by a constant value, see e.g. [godbolt here](https://godbolt.org/z/qKzYWWMWx). Both functions generate a right-shift by 4 in the final assembly.